File: src/main/java/com/ndnhuy/mybank/infra/QueueMetrics.java
package com.ndnhuy.mybank.infra;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Timer;
import lombok.Builder;
import lombok.Getter;

import java.util.concurrent.BlockingQueue;
import java.util.function.Supplier;

@Getter
public class QueueMetrics {
  private final MeterRegistry registry;

  // Arrival metrics
  private Counter transfersSubmitted;
  private Timer submissionTime;

  // Service metrics
  private Timer serviceTime;
  private Counter transfersCompleted;

  // Queue metrics (optional, can be null)
  private Gauge queueLength;
  private Timer waitTime;

  // System metrics (optional, can be null)
  private Gauge systemUtilization;

  // Time tracking for rate calculations
  private long startTime;

  private Runnable resetMetrics;

  public QueueMetrics(MeterRegistry registry) {
    this.registry = registry;
    this.startTime = System.nanoTime();
    Runnable initFunc = () -> {
      this.transfersSubmitted = Counter.builder("transfers.submitted").register(registry);
      this.submissionTime = Timer.builder("transfers.submission.time").register(registry);
      this.serviceTime = Timer.builder("transfers.service.time").register(registry);
      this.transfersCompleted = Counter.builder("transfers.completed").register(registry);
      this.waitTime = Timer.builder("transfers.wait.time").register(registry);
    };

    this.resetMetrics = () -> {
      // Remove existing meters from registry
      registry.remove(transfersSubmitted);
      registry.remove(submissionTime);
      registry.remove(serviceTime);
      registry.remove(transfersCompleted);
      registry.remove(waitTime);

      if (queueLength != null) {
        registry.remove(queueLength);
        queueLength = null;
      }

      if (systemUtilization != null) {
        registry.remove(systemUtilization);
        systemUtilization = null;
      }

      // Reset start time for new observation period
      this.startTime = System.nanoTime();

      initFunc.run();
    };

    initFunc.run();
  }

  public void setQueueLengthGauge(BlockingQueue<?> queue, MeterRegistry registry) {
    this.queueLength = Gauge.builder("transfers.queue.length", queue, q -> q.size()).register(registry);
  }

  public void setSystemUtilizationGauge(Supplier<Double> utilizationSupplier, MeterRegistry registry) {
    // Gauge.builder expects Supplier<Number>
    this.systemUtilization = Gauge.builder("system.utilization", (Supplier<Number>) utilizationSupplier::get)
        .register(registry);
  }

  public void reset() {
    if (resetMetrics != null) {
      resetMetrics.run();
    } else {
      throw new IllegalStateException("Reset function is not initialized");
    }
  }

  public Report getReport() {
    return Report.builder()
        .transfersSubmitted(transfersSubmitted.count())
        .submissionTimeInMilliseconds(submissionTime.totalTime(java.util.concurrent.TimeUnit.MILLISECONDS))
        .serviceTimeInMilliseconds(serviceTime.totalTime(java.util.concurrent.TimeUnit.MILLISECONDS))
        .transfersCompleted(transfersCompleted.count())
        .queueLength(queueLength != null ? queueLength.value() : 0)
        .waitTimeInMilliseconds(waitTime.totalTime(java.util.concurrent.TimeUnit.MILLISECONDS))
        .systemUtilization(systemUtilization != null ? systemUtilization.value() : 0)
        .queueMetrics(this)  // Pass reference to parent QueueMetrics
        .build();
  }

  // ============================================
  // Queuing Theory Statistical Methods
  // ============================================

  /**
   * Get the observation duration in seconds since metrics collection started.
   * @return observation duration in seconds
   */
  public double getObservationDurationSeconds() {
    return (System.nanoTime() - startTime) / 1_000_000_000.0;
  }

  /**
   * Get the mean response time (total time from arrival to completion).
   * Response Time = Wait Time + Service Time
   * @return mean response time in milliseconds, or 0 if no requests completed
   */
  public double getMeanResponseTime() {
    double completed = transfersCompleted.count();
    if (completed == 0) return 0;
    
    double totalWaitTime = waitTime.totalTime(java.util.concurrent.TimeUnit.MILLISECONDS);
    double totalServiceTime = serviceTime.totalTime(java.util.concurrent.TimeUnit.MILLISECONDS);
    return (totalWaitTime + totalServiceTime) / completed;
  }

  /**
   * Get the mean wait time (average time spent waiting in queue).
   * @return mean wait time in milliseconds, or 0 if no requests completed
   */
  public double getMeanWaitTime() {
    double completed = transfersCompleted.count();
    if (completed == 0) return 0;
    
    return waitTime.totalTime(java.util.concurrent.TimeUnit.MILLISECONDS) / completed;
  }

  /**
   * Get the mean service time (average time spent being processed).
   * @return mean service time in milliseconds, or 0 if no requests completed
   */
  public double getMeanServiceTime() {
    double completed = transfersCompleted.count();
    if (completed == 0) return 0;
    
    return serviceTime.totalTime(java.util.concurrent.TimeUnit.MILLISECONDS) / completed;
  }

  /**
   * Get the arrival rate (λ) - requests submitted per second.
   * @return arrival rate in requests/second
   */
  public double getArrivalRate() {
    double duration = getObservationDurationSeconds();
    if (duration <= 0) return 0;
    
    return transfersSubmitted.count() / duration;
  }

  /**
   * Get the service rate (μ) - requests completed per second.
   * Also known as throughput.
   * @return service rate in requests/second
   */
  public double getServiceRate() {
    double duration = getObservationDurationSeconds();
    if (duration <= 0) return 0;
    
    return transfersCompleted.count() / duration;
  }

  /**
   * Get the throughput (same as service rate).
   * @return throughput in requests/second
   */
  public double getThroughput() {
    return getServiceRate();
  }

  /**
   * Get the traffic intensity (ρ = λ/μ).
   * Represents the system load factor.
   * - ρ < 1: System is stable
   * - ρ = 1: System at capacity
   * - ρ > 1: System overloaded
   * @return traffic intensity (dimensionless)
   */
  public double getTrafficIntensity() {
    double serviceRate = getServiceRate();
    if (serviceRate <= 0) return Double.POSITIVE_INFINITY;
    
    return getArrivalRate() / serviceRate;
  }

  /**
   * Get the average queue length using Little's Law (L = λW).
   * L = average number of customers in the queue
   * λ = arrival rate
   * W = average wait time
   * @return average queue length (number of requests)
   */
  public double getAverageQueueLength() {
    double arrivalRate = getArrivalRate();
    double meanWaitTimeSeconds = getMeanWaitTime() / 1000.0; // Convert ms to seconds
    
    return arrivalRate * meanWaitTimeSeconds;
  }

  /**
   * Get the current instantaneous queue length.
   * @return current queue length (snapshot)
   */
  public double getCurrentQueueLength() {
    return queueLength != null ? queueLength.value() : 0;
  }

  /**
   * Get the current system utilization percentage.
   * @return system utilization as percentage (0-100)
   */
  public double getSystemUtilizationPercentage() {
    if (systemUtilization == null) return 0;
    return systemUtilization.value() * 100;
  }

  @Builder
  public static class Report {
    private final double transfersSubmitted;
    private final double submissionTimeInMilliseconds;
    private final double serviceTimeInMilliseconds;
    private final double transfersCompleted;
    private final double queueLength;
    private final double waitTimeInMilliseconds;
    private final double systemUtilization;
    private final QueueMetrics queueMetrics;

    public void print() {
      if (queueMetrics == null) {
        printLegacyReport();
        return;
      }

      System.out.println("╔══════════════════════════════════════════════════════════════╗");
      System.out.println("║                    QUEUING SYSTEM ANALYSIS                  ║");
      System.out.println("╚══════════════════════════════════════════════════════════════╝");
      
      // System Overview
      System.out.println("\n📊 SYSTEM OVERVIEW:");
      System.out.printf("   Observation Duration: %.2f seconds%n", queueMetrics.getObservationDurationSeconds());
      System.out.printf("   Requests Submitted:   %.0f%n", transfersSubmitted);
      System.out.printf("   Requests Completed:   %.0f%n", transfersCompleted);
      
      double trafficIntensity = queueMetrics.getTrafficIntensity();
      String systemStatus = getSystemStatus(trafficIntensity);
      System.out.printf("   System Status:        %s%n", systemStatus);

      // Performance Metrics
      System.out.println("\n⚡ PERFORMANCE METRICS:");
      System.out.printf("   Mean Response Time:   %.2f ms%n", queueMetrics.getMeanResponseTime());
      System.out.printf("   Mean Wait Time:       %.2f ms%n", queueMetrics.getMeanWaitTime());
      System.out.printf("   Mean Service Time:    %.2f ms%n", queueMetrics.getMeanServiceTime());
      System.out.printf("   Throughput:           %.2f requests/sec%n", queueMetrics.getThroughput());

      // Queuing Theory Analysis
      System.out.println("\n🔬 QUEUING THEORY ANALYSIS:");
      System.out.printf("   Arrival Rate (λ):      %.2f requests/sec%n", queueMetrics.getArrivalRate());
      System.out.printf("   Service Rate (μ):      %.2f requests/sec%n", queueMetrics.getServiceRate());
      System.out.printf("   Traffic Intensity (ρ): %.3f %s%n", trafficIntensity, getTrafficIntensityWarning(trafficIntensity));
      System.out.printf("   Avg Queue Length (L):  %.2f requests%n", queueMetrics.getAverageQueueLength());
      System.out.printf("   Current Queue Length:  %.0f requests%n", queueMetrics.getCurrentQueueLength());
      System.out.printf("   System Utilization:    %.1f%%%n", queueMetrics.getSystemUtilizationPercentage());

      // Performance Assessment
      System.out.println("\n🎯 PERFORMANCE ASSESSMENT:");
      printPerformanceAssessment(queueMetrics);
      
      System.out.println("\n" + "═".repeat(66));
    }

    private void printLegacyReport() {
      System.out.println("Queue Metrics Report (Legacy):");
      System.out.printf("Transfers Submitted: %.0f%n", transfersSubmitted);
      System.out.printf("Submission Time (ms): %.2f%n", submissionTimeInMilliseconds);
      System.out.printf("Service Time (ms): %.2f%n", serviceTimeInMilliseconds);
      System.out.printf("Transfers Completed: %.0f%n", transfersCompleted);
      System.out.printf("Queue Length: %.0f%n", queueLength);
      System.out.printf("Wait Time (ms): %.2f%n", waitTimeInMilliseconds);
      System.out.printf("System Utilization: %.2f%%%n", systemUtilization * 100);
    }

    private String getSystemStatus(double trafficIntensity) {
      if (trafficIntensity >= 1.0) return "🔴 OVERLOADED";
      if (trafficIntensity >= 0.8) return "🟡 HIGH LOAD";
      if (trafficIntensity >= 0.5) return "🟢 MODERATE LOAD";
      return "🟢 LOW LOAD";
    }

    private String getTrafficIntensityWarning(double trafficIntensity) {
      if (trafficIntensity >= 1.0) return "⚠️  SYSTEM OVERLOADED!";
      if (trafficIntensity >= 0.8) return "⚠️  Approaching capacity";
      return "✅ Stable";
    }

    private void printPerformanceAssessment(QueueMetrics metrics) {
      double responseTime = metrics.getMeanResponseTime();
      double waitTime = metrics.getMeanWaitTime();
      double trafficIntensity = metrics.getTrafficIntensity();

      System.out.printf("   Response Time:  %s%n", assessResponseTime(responseTime));
      System.out.printf("   Queue Buildup:  %s%n", assessQueueBuildup(waitTime));
      System.out.printf("   System Health:  %s%n", assessSystemHealth(trafficIntensity));
    }

    private String assessResponseTime(double responseTime) {
      if (responseTime < 100) return "🟢 Excellent (< 100ms)";
      if (responseTime < 500) return "🟡 Good (< 500ms)";
      if (responseTime < 1000) return "🟠 Fair (< 1s)";
      return "🔴 Poor (> 1s)";
    }

    private String assessQueueBuildup(double waitTime) {
      if (waitTime < 10) return "🟢 Minimal queue buildup";
      if (waitTime < 100) return "🟡 Moderate queue buildup";
      if (waitTime < 500) return "🟠 Significant queue buildup";
      return "🔴 Severe queue buildup";
    }

    private String assessSystemHealth(double trafficIntensity) {
      if (trafficIntensity >= 1.0) return "🔴 System cannot keep up with demand";
      if (trafficIntensity >= 0.8) return "🟡 System near capacity, consider scaling";
      if (trafficIntensity >= 0.5) return "🟢 System handling load well";
      return "🟢 System has excess capacity";
    }
  }
}

File: clienttest/loadtest/queue_metrics.go
package loadtest

import (
	"fmt"
	"strings"
	"time"

	vegeta "github.com/tsenart/vegeta/v12/lib"
)

// QueueMetrics wraps vegeta.Metrics with additional queuing theory calculations
type QueueMetrics struct {
	metrics   *vegeta.Metrics
	startTime time.Time
}

// NewQueueMetrics creates a new QueueMetrics instance
func NewQueueMetrics() *QueueMetrics {
	return &QueueMetrics{
		metrics:   &vegeta.Metrics{},
		startTime: time.Now(),
	}
}

// Close properly closes CSV file and cleans up resources
func (qm *QueueMetrics) Close() {
	// Close vegeta metrics to finalize the data
	qm.metrics.Close()
}

// GetArrivalRate returns the arrival rate (λ) in requests/second
func (qm *QueueMetrics) GetArrivalRate() float64 {
	return qm.metrics.Rate
}

// GetServiceRate returns the service rate (μ) in requests/second
func (qm *QueueMetrics) GetServiceRate() float64 {
	return qm.metrics.Throughput
}

// GetTrafficIntensity returns the traffic intensity (ρ = λ/μ)
func (qm *QueueMetrics) GetTrafficIntensity() float64 {
	if qm.metrics.Throughput <= 0 {
		return 999.0 // Indicate overload
	}
	return qm.metrics.Rate / qm.metrics.Throughput
}

// GetObservationDuration returns how long the test ran
func (qm *QueueMetrics) GetObservationDuration() time.Duration {
	return qm.metrics.Duration
}

// GetSystemStatus returns a color-coded status based on traffic intensity
func (qm *QueueMetrics) GetSystemStatus() string {
	rho := qm.GetTrafficIntensity()
	if rho >= 1.0 {
		return "🔴 OVERLOADED"
	}
	if rho >= 0.8 {
		return "🟡 HIGH LOAD"
	}
	if rho >= 0.5 {
		return "🟢 MODERATE LOAD"
	}
	return "🟢 LOW LOAD"
}

// AssessResponseTime provides response time assessment
func (qm *QueueMetrics) AssessResponseTime() string {
	mean := qm.metrics.Latencies.Mean
	if mean < 100*time.Millisecond {
		return "🟢 Excellent (< 100ms)"
	}
	if mean < 500*time.Millisecond {
		return "🟡 Good (< 500ms)"
	}
	if mean < 1*time.Second {
		return "🟠 Fair (< 1s)"
	}
	return "🔴 Poor (> 1s)"
}

// AssessSystemHealth provides overall system health assessment
func (qm *QueueMetrics) AssessSystemHealth() string {
	rho := qm.GetTrafficIntensity()
	if rho >= 1.0 {
		return "🔴 System cannot keep up with demand"
	}
	if rho >= 0.8 {
		return "🟡 System near capacity, consider scaling"
	}
	if rho >= 0.5 {
		return "🟢 System handling load well"
	}
	return "🟢 System has excess capacity"
}

// PrintReport prints a comprehensive report similar to Java QueueMetrics
func (qm *QueueMetrics) PrintReport() {
	fmt.Println("╔══════════════════════════════════════════════════════════════╗")
	fmt.Println("║                    QUEUING SYSTEM ANALYSIS                  ║")
	fmt.Println("╚══════════════════════════════════════════════════════════════╝")

	// System Overview
	fmt.Println("\n📊 SYSTEM OVERVIEW:")
	fmt.Printf("   Observation Duration: %v\n", qm.GetObservationDuration())
	fmt.Printf("   Requests Submitted:   %d\n", int(qm.metrics.Requests))
	fmt.Printf("   Requests Completed:   %d\n", int(qm.metrics.Requests)) // Assuming all completed for HTTP
	fmt.Printf("   System Status:        %s\n", qm.GetSystemStatus())

	// Performance Metrics
	fmt.Println("\n⚡ PERFORMANCE METRICS:")
	fmt.Printf("   Mean Response Time:   %v\n", qm.metrics.Latencies.Mean)
	fmt.Printf("   50th percentile:      %v\n", qm.metrics.Latencies.P50)
	fmt.Printf("   95th percentile:      %v\n", qm.metrics.Latencies.P95)
	fmt.Printf("   99th percentile:      %v\n", qm.metrics.Latencies.P99)
	fmt.Printf("   Max Response Time:    %v\n", qm.metrics.Latencies.Max)
	fmt.Printf("   Success Rate:         %.2f%%\n", qm.metrics.Success*100)

	// Queuing Theory Analysis
	fmt.Println("\n🔬 QUEUING THEORY ANALYSIS:")
	fmt.Printf("   Arrival Rate (λ):      %.2f requests/sec\n", qm.GetArrivalRate())
	fmt.Printf("   Service Rate (μ):      %.2f requests/sec\n", qm.GetServiceRate())
	fmt.Printf("   Traffic Intensity (ρ): %.3f\n", qm.GetTrafficIntensity())
	fmt.Printf("   Throughput:           %.2f requests/sec\n", qm.metrics.Throughput)

	// Performance Assessment
	fmt.Println("\n🎯 PERFORMANCE ASSESSMENT:")
	fmt.Printf("   Response Time:  %s\n", qm.AssessResponseTime())
	fmt.Printf("   System Health:  %s\n", qm.AssessSystemHealth())

	// Warnings
	if qm.metrics.Success < 1.0 {
		fmt.Printf("\n⚠️  Warning: Success rate is %.2f%%. Some requests failed!\n", qm.metrics.Success*100)
	}
	if qm.GetTrafficIntensity() >= 1.0 {
		fmt.Println("\n⚠️  SYSTEM OVERLOADED!")
	}

	fmt.Println("\n" + strings.Repeat("═", 66))
}

// Expose a public method to get a finalized snapshot of vegeta.Metrics
func (qm *QueueMetrics) Snapshot() *vegeta.Metrics {
	snapshot := qm.metrics
	snapshot.Close() // Finalize the snapshot for accurate percentiles
	return snapshot
}
